mava/systems/sac/anakin/ff_hasac.py

```diff
@@ -0,0 +1,728 @@
+    batch_size = obs.agents_view.shape[0]
+
+    actions = jnp.zeros((batch_size, num_agents, action_dim))
+    log_std = jnp.zeros((batch_size, num_agents))
+
+    for agent in range(num_agents):
+        actor_params_per_agent = tree.map(lambda x, agent=agent: x[agent], actor_params)
+        obs_per_agent = tree.map(lambda x, agent=agent: x[:, agent], obs)
+
+        pi = actor_net.apply(actor_params_per_agent, obs_per_agent)
+        action = pi.sample(seed=keys[agent])
+        actions = actions.at[:, agent].set(action)
+        log_std = log_std.at[:, agent].set(pi.log_prob(action))
+
+    return actions, log_std
+
+
+def init(
+    cfg: DictConfig,
+) -> Tuple[
+    Tuple[MarlEnv, MarlEnv],
+    Networks,
+    Optimisers,
+    TrajectoryBuffer,
+    LearnerState,
+    Array,
+    MavaLogger,
+    chex.PRNGKey,
+]:
+    """Initialize system by creating the envs, networks etc.
+
+    Args:
+    ----
+        cfg: System configuration.
+
+    Returns:
+    -------
+        Tuple containing:
+            Tuple[MarlEnv, MarlEnv]: The environment and evaluation environment.
+            Networks: Tuple of actor and critic networks.
+            Optimisers: Tuple of actor, critic and alpha optimisers.
+            TrajectoryBuffer: The replay buffer.
+            LearnerState: The initial learner state.
+            Array: The target entropy.
+            MavaLogger: The logger.
+            PRNGKey: The random key.
+    """
+    logger = MavaLogger(cfg)
+
+    key = jax.random.PRNGKey(cfg.system.seed)
+    devices = jax.devices()
+
+    def replicate(x: Any) -> Any:
+        """First replicate the update batch dim then put on devices."""
+        x = tree.map(lambda y: jnp.broadcast_to(y, (cfg.system.update_batch_size, *y.shape)), x)
+        return jax.device_put_replicated(x, devices)
+
+    env, eval_env = environments.make(cfg, add_global_state=True)
+
+    n_agents = env.num_agents
+    action_dim = env.action_dim
+
+    key, actor_key, q1_key, q2_key, q1_target_key, q2_target_key = jax.random.split(key, 6)
+    actor_keys = jax.random.split(actor_key, n_agents)
+
+    acts = env.action_spec().generate_value()  # all agents actions
+    act_single = acts[0]  # single agents action
+    concat_acts = jnp.concatenate([act_single for _ in range(n_agents)], axis=0)
+    concat_acts_batched = concat_acts[jnp.newaxis, ...]  # batch + concat of all agents actions
+    obs = env.observation_spec().generate_value()
+    obs_single_batched = tree.map(lambda x: x[0][jnp.newaxis, ...], obs)
+
+    # Making actor network
+    actor_torso = hydra.utils.instantiate(cfg.network.actor_network.pre_torso)
+    action_head, _ = get_action_head(env)
+    actor_action_head = hydra.utils.instantiate(
+        action_head, action_dim=env.action_dim, independent_std=False
+    )
+    actor_network = Actor(actor_torso, actor_action_head)
+    actor_params = jax.vmap(actor_network.init, in_axes=(0, None))(actor_keys, obs_single_batched)
```

Comment:
if you can add comment here to explain that the vmap used to produce different params per agent